src/Adapter/MSTest.TestAdapter/Execution/TestAssemblyInfo.cs

```diff
@@ -208,33 +213,46 @@ public void RunAssemblyInitialize(TestContext testContext)
             return null;
         }
 
+        Exception? assemblyCleanupException = null;
         lock (_assemblyInfoExecuteSyncObject)
         {
             try
             {
-                AssemblyCleanupMethod.InvokeAsSynchronousTask(null);
-
-                return null;
+                assemblyCleanupException = MethodRunner.RunWithTimeoutAndCancellation(
+                     () => AssemblyCleanupMethod.InvokeAsSynchronousTask(null),
+                     new CancellationTokenSource(),
+                     AssemblyCleanupMethodTimeoutMilliseconds,
+                     AssemblyCleanupMethod,
+                     Resource.AssemblyCleanupWasCancelled,
+                     Resource.AssemblyCleanupTimedOut);
             }
             catch (Exception ex)
             {
-                var realException = ex.GetRealException();
-
-                // special case AssertFailedException to trim off part of the stack trace
-                string errorMessage = realException is AssertFailedException or AssertInconclusiveException
-                    ? realException.Message
-                    : realException.GetFormattedExceptionMessage();
-
-                DebugEx.Assert(AssemblyCleanupMethod.DeclaringType?.Name is not null, "AssemblyCleanupMethod.DeclaringType.Name is null");
-                return string.Format(
-                    CultureInfo.CurrentCulture,
-                    Resource.UTA_AssemblyCleanupMethodWasUnsuccesful,
-                    AssemblyCleanupMethod.DeclaringType.Name,
-                    AssemblyCleanupMethod.Name,
-                    errorMessage,
-                    realException.GetStackTraceInformation()?.ErrorStackTrace);
+                assemblyCleanupException = ex;
             }
         }
+
+        // If assemblyCleanup was successful, then don't do anything
+        if (assemblyCleanupException is null)
+        {
+            return null;
+        }
+
+        var realException = assemblyCleanupException.InnerException ?? assemblyCleanupException;
```

Comment:
This change is removing the helper I added to avoid the type initialization exception.

```suggestion
        var realException = ex.GetRealException();
```